src/Errata/Internal/Render.hs

```diff
@@ -36,50 +38,110 @@ renderErrors :: Source source => source -> [Errata] -> TB.Builder
 renderErrors source errs = unsplit "\n\n" prettified
     where
         sortedErrata = sortOn (blockLocation . errataBlock) errs
-        slines = let xs = sourceToLines source in A.listArray (0, length xs - 1) xs
+        slines = sourceToLines source
         prettified = map (renderErrata slines) sortedErrata
 
 -- | A single pretty error from metadata and source lines.
-renderErrata :: Source source => A.Array Int source -> Errata -> TB.Builder
+renderErrata :: Source source => [source] -> Errata -> TB.Builder
 renderErrata slines (Errata {..}) = errorMessage
     where
+        -- | Header block + body blocks
+        blocks = errataBlock : errataBlocks
+
+        -- The pointers grouped by line.
+        blockPointersGrouped = I.fromListWith (<>) . map (pointerLine &&& pure) . blockPointers <$> blocks
+
+        -- Min and max line numbers as defined by the pointers of each block.
+        minPointers = maybe 1 id . fmap fst . I.lookupMin <$> blockPointersGrouped
+        maxPointers = maybe 0 id . fmap fst . I.lookupMax <$> blockPointersGrouped
+
+        minLine = minimum minPointers
+        maxLine = maximum maxPointers
+
+        -- Turns a list into a list of tail slices of the original list,
+        -- each element at index @i@ dropping the first @i@ elements of the original list.
+        slices [] = repeat (repeat mempty)
+        slices xs = (xs <> repeat mempty) : slices (tail xs)
```

Comment:
Are the `mempty`s here just `[]`? Why the need to repeat infinitely? I'm guessing to allow out-of-bounds access? Do add a comment and a type signature.